README.md

```diff
@@ -1 +1,144 @@
+- `fail(error)` - Creates an error result, specifically `{success: false, error: error}`
+
+and the following Typescript types:
+
+- `GoResult<T> = { data: T; success: true }`
+- `GoResultSuccess<E extends Error = Error> = { error: E; success: false }`
+- `GoResultError<T, E extends Error = Error> = GoResultSuccess<T> | GoResultError<E>`
+
+Take a look at the [implementation](https://github.com/api3dao/promise-utils/blob/main/src/index.ts) and
+[tests](https://github.com/api3dao/promise-utils/blob/main/src/index.test.ts) for detailed examples and usage.
+
+## Motivation
+
+### Verbosity and interoperability of try-catch pattern
+
+```ts
+// Verbose try catch
+try {
+  const data = await someAsyncCall();
+  ...
+} catch (e) {
+  // The "e" is "unknown" because any value can be thrown in Javascript so casting is needed
+  return logError((e as MyError).reason);
+}
+
+// Compare it to simpler version using go
+const goRes = await go<MyData, MyError>(someAsyncCall);
+if (!goRes.success) return logError(goRes.error.reason);
+// At this point TypeScript infers that the error was handled and "goRes" must be a success response
+const data = goRes.data;
+...
+```
+
+Also, think about what happens when you want to handle multiple "can fail" operations in a single function call. You can
+either:
+
+1. Have them in a same try catch block - but then it's difficult to differentiate between what error has been thrown.
+   Also this usually leads to a lot of code inside a try block and the catch clause acts more like "catch anything".
+2. Use nested try catch blocks - but this hurts readability and forces you into
+   [callback hell pattern](http://callbackhell.com/).
+
+### Consistent throwing of an `Error` instance
+
+Javascript supports throwing any expression, not just `Error` instances. This is also a reason why Typescript infers the
+error as `unknown` or `any` (see:
+[useUnknownInCatchVariables](https://www.typescriptlang.org/tsconfig#useUnknownInCatchVariables)).
+
+The error response from `go` and `goSync` always return an instance of the `Error` class. Of course, throwing custom
+errors (derived from `Error`) is supported.
+
+## Limitations
+
+There is a limitation when using class functions due to how javascript
+[this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this) works.
+
+```ts
+class MyClass {
+  constructor() {}
+  get() {
+    return this._get();
+  }
+  _get() {
+    return '123';
+  }
+}
+
+const myClass = new MyClass();
+const resWorks = goSync(() => myClass.get()); // This works
+// However, seeing the line above it may be tempting to rewrite it to
+const resFails = goSync(myClass.get); // This doesn't work
+```
+
+The problem is that the `this` keyword is determined by how a function is called and in the second example, the `this`
+insode the `get` function is `undefined` which makes the `this._get()` throw an error.
```

Comment:
```suggestion
inside the `get` function is `undefined` which makes the `this._get()` throw an error.
```

Reply:
I honestly re-read it after myself but didn't notice these (and my grammar extension in VScode either...)